File: problems/lc0046_Permutations.go
package problems

func Permute(nums []int) [][]int {
	result := make([][]int, 0)

	swapTwoElementByIndex := func(pList *[]int, i int, j int) {
		(*pList)[i], (*pList)[j] = (*pList)[j], (*pList)[i]
	}

	var backTrace func(fixedFirstNElement int)
	backTrace = func(fixedFirstNElement int) {
		if fixedFirstNElement == len(nums) {
			oneCombination := make([]int, len(nums), len(nums))
			copy(oneCombination, nums)
			result = append(result, oneCombination)
		}
		for i := fixedFirstNElement; i < len(nums); i++ {
			swapTwoElementByIndex(&nums, fixedFirstNElement, i)
			backTrace(fixedFirstNElement + 1)
			swapTwoElementByIndex(&nums, fixedFirstNElement, i)
		}
	}

	backTrace(0)
	return result
}

File: problems/Backtracking/lc0046_Permutations.py
def permute(nums: list[int]) -> list[list[int]]:
	result: list[list[int]] = []

	def swap_two_elements_by_index(list_: list[int], i: int, j: int) -> None:
		list_[i], list_[j] = list_[j], list_[i]

	def back_trace(fixed_first_n_elements: int) -> None:
		if fixed_first_n_elements == len(nums):
			result.append(nums.copy())
			return
		for unfixed_element_index in range(fixed_first_n_elements, len(nums)):
			swap_two_elements_by_index(nums, fixed_first_n_elements, unfixed_element_index)
			back_trace(fixed_first_n_elements + 1)
			swap_two_elements_by_index(nums, unfixed_element_index, fixed_first_n_elements)

	back_trace(0)
	return result
